src/app/server/CommissioningWindowManager.cpp

```diff
@@ -98,6 +98,10 @@ void CommissioningWindowManager::ResetState()
     mECMIterations    = 0;
     mECMSaltLength    = 0;
 
+#if CHIP_DEVICE_CONFIG_ENABLE_SED
+    DeviceLayer::ConnectivityMgr().RequestSEDActiveMode(false);
```

Comment:
Given that each `RequestSEDActiveMode` increases or decreases the SED active mode user counter, we should be very careful not to do it twice in either direction. Previously this was closely related to `mListeningForPASE` variable, so there was no such risk, but now we should perhaps add another member to keep track whether active mode has already been requested or not. Note that `AdvertiseAndListenForPASE` can be called multiple times in the case of unsuccessful pairing.

Reply:
I just saw the logic with the SED active counter and I'm not sure I'm fully in agreement with that. Putting the burden of doing a cleanup on the caller instead of having some protection directly inside the API could lead to some extreme corner case and prevent the device from ever going back to sleep.

e.g.
```
for (auto i = 0; i < 255 ; i++)
{
    DeviceLayer::ConnectivityMgr().RequestSEDActiveMode(true);
}

DeviceLayer::ConnectivityMgr().RequestSEDActiveMode(false);

```

I would have hope that the latest call will superseded the 255 previous one and put the device back to Idle mode, but given the current logic, we would have to repeat the exact same number of call to return back to Idle mode.

But Good point, I'll add a flag to prevent this in the `CommissioningWindowManager` while we think of a way to make the API more bullet proof (maybe there is no other way).

Reply:
Yeah, but `RequestSEDActiveMode` is also used by exchanges and the device should stay active if there's **at least** one user of the active mode, so if `RequestSEDActiveMode(false)`  superseded the previous 255 calls, it wouldn't work correctly for the exchange case. We could probably reserve some bit for commissioning but still the counter is needed for other use cases (unless we want to keep track which exchanges have requested the active mode inside the ThreadStackManager but that seems less memory-efficient).

Reply:
Yeah, that's an excellent point but another issue we also found is that there is no time factor taken into account. Like we cannot configure a SED to stay in Active mode for at least X amount of time or in Idle Mode for X amount of time. A possible solution would be to add `RequestSEDActiveMode(true, delay, increase)` With a timer inside the ThreadStackManager, that way the SED could revert back to Idle by itself. Calling explicitly the Idle mode would stop everything.

The way I see it, we could have a device, that upon reaching 10% battery capacity shutdown almost everything and Go in "hibernation" to preserve its core functionality. However there is no mechanism (at least that I'm aware of) that would grant the App layer with FULL control over the sleep time

Reply:
The time factor can be implemented in the existing `RequestSEDActiveMode` function - once the counter gets back to 0, start the timer which actually switches the underlying SED mode. `RequestSEDActiveMode(true)` cancels the timer.

When it comes to low-battery mode - couldn't the application simply call `ThreadStackManager::SetSEDIntervalsConfig` to increase both active and idle intervals in such a case?

Reply:
True, will look into it in a near future